### ISSUE TITLE
LTFS Formatting fails on Quantum LTO-6 Full-Height FC Drive (HP Ultrium 6-SCSI)

### ISSUE THREAD

**Issue by Nelyssia** [LinearTapeFileSystem/ltfs#490]
I am experiencing issues with LTFS on a Quantum LTO-6 Full-Height (FH) Fibre Channel tape drive (HP Ultrium 6-SCSI). The formatting process fails with the errors End-of-Data (EOD) not found and Invalid Field in Parameter List.

Could this issue stem from differences between Full-Height (FH) and Half-Height (HH) Quantum drives?

**Environment**

    Operating System: Ubuntu 22.04.5 LTS
    Kernel Version: 5.15.0-127-generic
    LTFS Version: 2.5.0.0 (Prelim)
    Tape Drive:
        Vendor: HP
        Product: Ultrium 6-SCSI
        Firmware: J5WZ
        Serial Number: HU1302U4Y2
        Quantum P/N: 1-04252-01
        HP Model: AQ278B#901
    HBA:
        QLogic QLE2460 4Gb Fibre Channel HBA
        PCI ID: 1077:2432
        Kernel driver in use: qla2xxx

**To Reproduce**

1. Compile LTFS with the following configuration: `./configure --enable-buggy-ifs
make
sudo make install`
2. Load a new Fujifilm LTO-6 BaFe tape into the drive.
3. Attempt to format the tape with LTFS: `mkltfs -d HU1302U4Y2 -t`
4. Observe the errors.

**Observed Behavior**
The formatting process fails with the following errors:
`End-of-Data (EOD) not found (-20802)`
`Invalid Field in Parameter List (-20500)`

Detailed Log:
```
mkltfs -d HU1302U4Y2 -t
LTFS15000I Starting mkltfs, LTFS version 2.5.0.0 (Prelim), log level 3.
LTFS15041I Launched by "mkltfs -d HU1302U4Y2 -t".
LTFS15042I This binary is built for Linux (x86_64).
LTFS15043I GCC version is 11.4.0.
LTFS17087I Kernel version: Linux version 5.15.0-127-generic (buildd@lcy02-amd64-019) (gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0, GNU ld (GNU Binutils for Ubuntu) 2.38) #137-Ubuntu SMP Fri Nov 8 15:21:01 UTC 2024 i386.
LTFS17089I Distribution: PRETTY_NAME="Ubuntu 22.04.5 LTS".
LTFS17089I Distribution: DISTRIB_ID=Ubuntu.
LTFS15025D Validating command line options.
LTFS15037D Command line options are valid.
LTFS15003I Formatting device 'HU1302U4Y2'.
LTFS15004I LTFS volume blocksize: 524288.
LTFS15005I Index partition placement policy: None.

LTFS11337I Update index-dirty flag (1) - NO_BARCODE (0x0x55c82c709cc0).
LTFS15006D Opening the device.
LTFS17085I Plugin: Loading "sg" tape backend.
LTFS30209I Opening a device through sg-ibmtape driver (HU1302U4Y2).
LTFS30288I Opening a tape device for drive serial HU1302U4Y2.
LTFS30250I Opened the SCSI tape device 7.0.0.0 (/dev/sg4).
LTFS30207I Vendor ID is HP      .
LTFS30208I Product ID is Ultrium 6-SCSI  .
LTFS30214I Firmware revision is J5WZ.
LTFS30215I Drive serial is HU1302U4Y2.
LTFS30290I Changer /dev/sg4 isn't reserved from any nodes.
LTFS30285I The reserved buffer size of HU1302U4Y2 is 1048576.
LTFS30294I Setting up timeout values from RSOC.
LTFS12023D Reserving device.
LTFS30392D Backend reserve (PRO) HU1302U4Y2.
LTFS12028D Unlocking medium.
LTFS30392D Backend allow medium removal HU1302U4Y2.
LTFS30392D Backend read block limits HU1302U4Y2.
LTFS17160I Maximum device block size is 1048576.
LTFS11330I Loading cartridge.
LTFS30392D Backend load HU1302U4Y2.
LTFS30398D Backend readpos: (0, 0) FM = 0 HU1302U4Y2.
LTFS12026D Locking medium in the drive.
LTFS30392D Backend prevent medium removal HU1302U4Y2.
LTFS30398D Backend readpos: (0, 0) FM = 0 HU1302U4Y2.
LTFS30392D Backend read block limits HU1302U4Y2.
LTFS11332I Load successful.
LTFS17157I Changing the drive setting to write-anywhere mode.
LTFS15007D Device opened.
LTFS15049I Checking the medium (mount).
LTFS11012D Loading the tape.
LTFS30392D Backend load HU1302U4Y2.
LTFS30398D Backend readpos: (0, 0) FM = 0 HU1302U4Y2.
LTFS30398D Backend readpos: (0, 0) FM = 0 HU1302U4Y2.
LTFS30392D Backend read block limits HU1302U4Y2.
LTFS30397D Backend locate: (0, 0) HU1302U4Y2.
LTFS30201D CDB check condition: sense = 081403, End-of-Data (EOD) not found.
LTFS30205I LOCATE (0x92) returns -20802.
LTFS30263I LOCATE returns End-of-Data (EOD) not found (-20802) /dev/sg4.
LTFS30398D Backend readpos: (0, 0) FM = 0 HU1302U4Y2.
LTFS12037E Cannot seek: backend call failed (-20802).
LTFS11006E Cannot read volume: failed to load the tape.
LTFS12028D Unlocking medium.
LTFS30392D Backend allow medium removal HU1302U4Y2.
LTFS12025D Releasing device.
LTFS30392D Backend release (PRO) HU1302U4Y2.
LTFS30393D Backend LBP Enable: 0 .
LTFS30393D Backend LBP Method: 1 .
LTFS30201D CDB check condition: sense = 052600, Invalid Field in Parameter List.
LTFS30205I MODESELECT (0x55) returns -20500.
LTFS30263I MODESELECT returns Invalid Field in Parameter List (-20500) /dev/sg4.
LTFS15023I Formatting failed.
```

**Expected Behavior**
The tape should be formatted successfully using LTFS.

**System Details**

- Operating System: Ubuntu 22.04.5 LTS
- Kernel Version: 5.15.0-127-generic
- LTFS Version: 2.5.0.0 (Prelim)

**Hardware:
Tape Drive:**
sg_inq /dev/sg4:
`Vendor identification: HP
Product identification: Ultrium 6-SCSI
Product revision level: J5WZ
Unit serial number: HU1302U4Y2`
lsscsi -l:
`[7:0:0:0]    tape    HP       Ultrium 6-SCSI   J5WZ  /dev/st0
  state=running queue_depth=64 scsi_level=7 type=1 device_blocked=0 timeout=900`

**HBA:**
`02:00.0 Fibre Channel: QLogic Corp. ISP2432-based 4Gb Fibre Channel to PCI Express HBA (rev 03)
Subsystem: QLogic Corp. QLE2460 Single-Port 4Gbps FC-to-PCI-X/PCIe Host Bus Adapter
Kernel driver in use: qla2xxx`

**What I’ve Tried**
- Using multiple Fujifilm LTO-6 BaFe tapes.
- Enabling --enable-buggy-ifs during the LTFS build process.
- Testing with freshly loaded tapes.

**Comment by piste-jp** [comment]
The key of this error is below. It means LTFS issues the `LOCATE pos=0, part=0` command but the drives returns an error, check condition with `nd-of-Data (EOD) not found`.

```
LTFS30397D Backend locate: (0, 0) HU1302U4Y2.
LTFS30201D CDB check condition: sense = 081403, End-of-Data (EOD) not found.
LTFS30205I LOCATE (0x92) returns -20802.
LTFS30263I LOCATE returns End-of-Data (EOD) not found (-20802) /dev/sg4.
```

It looks the tape was not unloaded from the drive cleanly in the previous mount, like sudden power loss. Anyway, the drive shall not report this kind of check condition against `LOCATE 0, 0` or shell recover this condition automatically.

May be `ERASE` command just after cartridge load might recover the condition, I guess. You might have a value to try that with `mt` command like `mt /dev/st? erase`.

**Comment by trosenkranz** [comment]
TLDR: Try to format your tape with `mkltfs` AND the `-f` (force) command line option.

@piste-jp I recently had a similar issue with `mkltfs` and my HP LTO-6 tape drive. I could not format a brand new tape. After fiddling around a bit I wrote a file to the tape using the `tar` command. After that the formatting with `mkltfs` worked fine.

After a good nights sleep it hit me. I recreated the situation with another new tape. I got the exact same error message you got. It essentially says, that there has been never written any data to the tape before. When I executed the `mkltfs` command with the `-f` parameter it worked fine.

I strongly advise against using the force option per default because it also prevents you of accidentally overwriting an existing _ltfs_ filesystem on the tape.

I don't know how the `mt ... erase ` command exactly works. According to the manpage it overwrites the tape completely (with zeros i assume), but i don't know whether it also writes an _EOD_ marker, which is necessary to format a tape without the force option. Have you tried that? Did it work?

I hope it helps and let me know whether it work.

Have a nice day
Tobias

**Comment by piste-jp** [comment]
> I don't know how the `mt ... erase ` command exactly works. According to the manpage it overwrites the tape completely (with zeros i assume), but i don't know whether it also writes an _EOD_ marker, which is necessary to format a tape without the force option. Have you tried that? Did it work?

Actually, `mt` command just sends a `ERASE` command to the tape drive. 

Basically, there are 2 types of erase on the tape drive. 

1. Long erase
2. Short erase

In the long erase, tape drive might write unmeaning data from the current position to the end of tape. After that tape drive back to the position which the erase command is issued, and then write an EOD marker.

In the short erase, tape drive just write an EOD marker in the current position.

---

In the man page, you can issue short erase with count=0. The point here is erase command doesn't mean "erase all data on tape", but "erase all data after current position". So you need to set position 0 (as rewind command) if you want to erase all data on tape.

```
erase  Perform long erase of tape. If count is 0, perform short erase of tape (some devices do not support this).
```

**Comment by piste-jp** [comment]
Close this because long time no activity.
